TACHO_DRIVER.md

```diff
@@ -0,0 +1,199 @@
+- [ ] [ev3dev/ev3dev#224](https://github.com/ev3dev/ev3dev/issues/224)
+- [ ] [ev3dev/ev3dev#149](https://github.com/ev3dev/ev3dev/issues/149)
+- [ ] [ev3dev/ev3dev#148](https://github.com/ev3dev/ev3dev/issues/148)
+- [ ] [ev3dev/ev3dev#54](https://github.com/ev3dev/ev3dev/issues/54)
+
+Other tasks that need to be dealt with for this proposal:
+
+- [ ] Add LibreOffice Template for graphs - I'd like to make sure that we have
+a consistent format for the diagrams so that when they need changes we're not
+recreating them from scratch.
+- [ ] Add pointers to spec.json here so that we have a consistent notation
+for commands, functions, and mode. We may need to update `spec.json` so that
+the operation is better specified there.
+
+In order to properly fix the issues, and add the `sync` feature that users have
+been asking for, this document describes the operating modes and functions for
+the tacho motors.
+
+The `tacho-motor` drivers must strike a balance between doing the most common
+functions that a robot builder needs but not becoming so complex that the
+operation is unreliable or mysterious.
+
+## The sys Interface
+
+All motor operations are contolled by writing data to attributes within the
+`/sys/class/tacho-motor/motorN/` folder. There is one folder per motor.
+
+Some of the attributes are valid only during particular `run` modes - and with
+that in mind, the read and write operations can have the following return codes:
+
+| RETURN CODE  | DESCRIPTION
+| ------------ | -----------
+| 0            | Operation sucessful
+| ( -1) EPERM  | Operation not permitted (not the same as EACCESS!) 
+| (-13) EACCES | Permission denied
+| (-13) EBUSY  | Device or resource busy
+| (-22) EINVAL | Invalid argument
+
+The attributes operate as described in `spec.json` - with the following
+caveats for the `speed_regulation` value:
+
+| `speed_regulation` | Write `duty_cycle_sp` | Write `speed_sp`
+| ------------------ | --------------------- | ----------------
+| `off`              | Works                 | Return EPERM
+| `on`               | Return EPERM          | Works
+
+The `ramp_up_sp` and `ramp_down_sp` specify how long (in msec) the motor
+will take to go from 0 to 100 of the controlling parameter depending on
+the `speed_regulation` mode.
+
+| `speed_regulation` | Ramp control parameter
+| ------------------ | ----------------------
+| `off`              | 0 to 100% Duty Cycle                 
+| `on`               | 0 to 100% Max Speed
+
+The maximum speed of the motor is set in the driver. The large motor has a
+maximum speed of 1200 pulses per second and the medium motor has a maximum
+speed of 900 pulses per second. Both motors have 360 pulses per revolution.
+
+## Ramping
+
+See `spec.json` for a description of the `ramp_up_sp` and `ramp_down_sp`
+attributes. Except in `run-direct` mode, you cannot change the ramping
+attributes while the motor is running.
+
+If you change the allowed `duty_cycle_sp` or `speed_sp` value while the
```

Comment:
I'll push back on this - when `speed_regulation` is `on` I would argue that we DO want to be able to change `speed_sp` on the fly in `run_direct` mode. I'm not sure how it's any harder to implement.

It would have the benefit of allowing users to create non-linear ramping if they need it. The kernel module supports linear ramping - if you need non-linear, use `run_direct` and write it yourself.